pythonPackages.django_3: init at 3

FRidh opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

FRidh committed:
pythonPackages.django_3: init at 3.1

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -3750,12 +3750,14 @@ in {
   crayons = callPackage ../development/python-modules/crayons{ };
 
-  django = self.django_lts;
+  django = self.django_3_1;
```
It might be wiser to just keep this the same as the lts version, that way we will have fewer breakage I expect. Anyway, I think we need to choose what versions of the packages we focus on, those compatible with the latest django, or those compatible with lts django. I prefer the latter because I think it becomes a mess otherwise..

lsix replied:
Hi.

I really think it makes the maintenance easier if the “main” django version we support is the LTS one. I initially added the `django_lts` alias to make it clear that `django` is the current LTS and not the latest available version.

I think we could however introduce a `django_latest` or something like this if we wish to offer a more up to date alternative.

An alternative (that would require a fair bit of rework) would be to propose a django subset within the python package set, and instantiate such subset for multiple django versions. I am not sure if going this way is worth the effort but can give a try (later) if you think this could be beneficial.

FRidh replied:
No, let's not add more attributes, and let's remove `lts`. `django` will then always point to `lts`. The latest version we can add like I did here, but without aliasing it.

FRidh replied:
Updated this PR.

jonringer commented:
there's a lot of packages with similar breakages:
```
  ERROR: Could not find a version that satisfies the requirement Django<3.1,>=1.11 (from django-mailman3==1.3.4) (from versions: none)
  ERROR: No matching distribution found for Django<3.1,>=1.11 (from django-mailman3==1.3.4)
```

FRidh committed:
pythonPackages.django: clarify versions, remove django_lts attribute

Every major version will have only a single LTS release, which is the
one we typically keep in Nixpkgs.

https://www.djangoproject.com/download/

lsix commented on pkgs/development/python-modules/django/3.nix:
```diff
@@ -0,0 +1,49 @@
+
+  patches = stdenv.lib.optional withGdal
+    (substituteAll {
+      src = ./1.10-gis-libs.template.patch;
```
This patch is not valid for django-3.

You can re-use the one I prepared when I did a django-3.0 PR (#92848)

```patch
diff -Nur a/django/contrib/gis/gdal/libgdal.py b/django/contrib/gis/gdal/libgdal.py
--- a/django/contrib/gis/gdal/libgdal.py	2020-07-09 22:34:05.330568948 +0100
+++ b/django/contrib/gis/gdal/libgdal.py	2020-07-09 22:35:08.679095615 +0100
@@ -14,7 +14,7 @@
     from django.conf import settings
     lib_path = settings.GDAL_LIBRARY_PATH
 except (AttributeError, ImportError, ImproperlyConfigured, OSError):
-    lib_path = None
+    lib_path = "@gdal@/lib/libgdal@extension@"

 if lib_path:
     lib_names = None
diff -Nur a/django/contrib/gis/geos/libgeos.py b/django/contrib/gis/geos/libgeos.py
--- a/django/contrib/gis/geos/libgeos.py	2020-07-09 22:34:05.331568941 +0100
+++ b/django/contrib/gis/geos/libgeos.py	2020-07-09 22:36:24.863526276 +0100
@@ -24,7 +24,7 @@
         from django.conf import settings
         lib_path = settings.GEOS_LIBRARY_PATH
     except (AttributeError, ImportError, ImproperlyConfigured, OSError):
-        lib_path = None
+        lib_path = "@geos@/lib/libgeos_c@extension@"

     # Setting the appropriate names for the GEOS-C library.
     if lib_path:
```

lsix reviewed (changes requested):
Hi,

Sorry for the delay for this review (I am currently in holiday, mostly AFK).

Overall, I am OK with the PR. Just the GIS support is not valid as is, the patch used to indicate where to find libgdal / libgeos is not valid anymore for django-3 branch.

I gave you a patch that can be used instead.

Also there seems to be merge conflicts in `pkgs/top-level/python-packages.nix`. This PR should probably be rebased to master.

jonringer commented:
please rebase on the latest master. I sorted `python-packages.nix`

lsix commented:
@FRidh I am back from my break. Have you had time to rework this PR, or do you want me to take over it?

FRidh commented:
@lsix feel free to take it over!

lsix commented:
I have updated the work done here in #98463

I close this PR in favor of the new one.
